pkg/phlaredb/bucket/tenant_scanner.go

```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cortexproject/cortex/blob/master/pkg/storage/tsdb/users_scanner.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Cortex Authors.
+
+package bucket
+
+import (
+	"context"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+
+	"github.com/grafana/pyroscope/pkg/objstore"
+)
+
+// AllTenants returns true to each call and should be used whenever the UsersScanner should not filter out
+// any user due to sharding.
+func AllTenants(_ string) (bool, error) {
+	return true, nil
+}
+
+type TenantsScanner struct {
+	bucketClient objstore.Bucket
+	logger       log.Logger
+	isOwned      func(userID string) (bool, error)
+}
+
+func NewTenantsScanner(bucketClient objstore.Bucket, isOwned func(userID string) (bool, error), logger log.Logger) *TenantsScanner {
+	return &TenantsScanner{
+		bucketClient: bucketClient,
+		logger:       logger,
+		isOwned:      isOwned,
+	}
+}
+
+// ScanTenants returns a fresh list of users found in the storage, that are not marked for deletion,
+// and list of users marked for deletion.
+//
+// If sharding is enabled, returned lists contains only the users owned by this instance.
+func (s *TenantsScanner) ScanTenants(ctx context.Context) (users, markedForDeletion []string, err error) {
```

Comment:
I'm unsure if this is always true. As far as I can see we may return users that are not owned by this instance if `isOwned` fails. I'm pretty sure this is intended behaviour, judging by the tests. It should be okay only if the tenant scanner consumers can handle the same users concurrently (e.g if the only use case is removing tenants). @cyriltovena are there other use cases?

Reply:
Yes I didn't realize this I'll check the usage of the scanner, I think it was just for cleaning blocks and tenants.

Reply:
Seems to only be used for cleaning yes, so we should be fine, nice catch though